### ISSUE TITLE
Partial Success, but can't go past December 2016

### ISSUE THREAD

**Issue by TheVeeC** [drJeckyll/tumblr-chat-messages-downloader#4]
I've tried on two separate days and though it works just fine, it's also failing after going back a certain amount of time.

As shown in a snippet of the results below (since the large list of successful dates/times is irrelevant):
10/12/2016, 04:17:19, 10/12/2016, 04:15:34, 10/12/2016, 03:43:01, 10/12/2016, 03:29:17, 10/12/2016, 03:24:22, 10/12/2016, 03:09:26, 10/12/2016, 02:56:54, 09/12/2016, 22:23:25, 09/12/2016, 22:00:28, 09/12/2016, 21:52:47, retry, retry, retry, retry, retry, PHP Notice:  Trying to get property 'response' of non-object in C:\xampp\tumblr\tumblr.php on line 157

Notice: Trying to get property 'response' of non-object in C:\xampp\tumblr\tumblr.php on line 157
PHP Notice:  Trying to get property 'messages' of non-object in C:\xampp\tumblr\tumblr.php on line 157

Notice: Trying to get property 'messages' of non-object in C:\xampp\tumblr\tumblr.php on line 157
PHP Notice:  Trying to get property 'data' of non-object in C:\xampp\tumblr\tumblr.php on line 157

Notice: Trying to get property 'data' of non-object in C:\xampp\tumblr\tumblr.php on line 157
PHP Warning:  count(): Parameter must be an array or an object that implements Countable in C:\xampp\tumblr\tumblr.php on line 159

Warning: count(): Parameter must be an array or an object that implements Countable in C:\xampp\tumblr\tumblr.php on line 159
done.

I know for a fact that there is more to the conversation history than this.  I've even been able to scroll up past this in tumblr, after about an hour, just to confirm that the logs do go back further.

I've tried downloading everything in one document, using -s to split into multiple documents and I've even tried using -d to try earlier dates; all attempts have had the exact same result.  Apparently even -d still tries to go through the entire list of dates before failing to reach the requested date and creating a blank document in my case.

Any ideas on how to fix this problem and go back further?

**Comment by drJeckyll** [comment]
There is data for sure but for some reason script fail to get it. 

These "retry" messages are printed when returned data is not a valid json object. 

You can try to add: 
`echo "received: " . $r . "\n";` 
after `$r = curl_exec($ch);` on row 150 and before `$r = json_decode($r);`. 

This must print what you receive from tumblr and give us a hint what is wrong.

`-s`, `-d` are just for format on client side. Since tumblr don't provide API for chat, there is no know method (at least for me) to get messages for certain date(s). So script just emulate user scrolling back through chat history. On every scroll tumblr return chat data along with identifier for previous data. Script just follows these identifiers just like when user go up though his chat history. I know it sucks, but that's what we have.

**Comment by TheVeeC** [comment]
Apparently tumblr has a rate limit.  This repeated numerous times:

09/12/2016, 21:52:47, received: <!DOCTYPE html>

<html>
    <head>
        <title>Rate limit exceeded.</title>
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
        <meta http-equiv="x-dns-prefetch-control" content="off">
        <meta name="robots" content="noindex">
        <meta name="viewport" content="width=device-width">
        <meta name="application-name" content="Tumblr">
        <meta name="msapplication-TileColor" content="#2c4762">
        <meta name="msapplication-TileImage" content="//assets.tumblr.com/images/msfavicon.png?_v=245323c5cb69e705ea213d9ed60e543a">
        <link rel="shortcut icon" href="//assets.tumblr.com/images/favicons/favicon.ico?_v=8bfa6dd3e1249cd567350c606f8574dc">
        <link rel="stylesheet" media="screen" href="//assets.tumblr.com/client/prod/standalone/error-pages/index.build.css?_v=da302762836d5bf2aeee0f6522476e59">

    </head>

    <body data-status-code="429">
        <div class="error-message-container">
            <div class="error-message-content">
                <h1 class="error-message-title" data-localization="title">
                    Rate limit exceeded.
                </h1>

                <p class="error-message-text" data-localization="sorry">
                    <a href="https://www.tumblr.com/docs/error_troubleshooting#rate_limit_exceeded" target="_docs">Find out why</a> you may have encountered this error, or let us know if the problem persists. Include your IP address and a short description of what you were doing when you encountered the rate limit.
            </p>
            </div>
        </div>

        <script src="//assets.tumblr.com/languages/errors.js?_v=aa8ebc5ca20b127a3a667152c15432a2"></script>
        <script src="//assets.tumblr.com/client/prod/standalone/error-pages/index.build.js?_v=f30b986ee7cbce9d3a780ec26d1e7514"></script>

    </body>
</html>

Each retry fetched the same thing.

Seems like a timer between requests should fix the issue, even if it would take longer.  I'm not terribly familiar with PHP though, so I'm not sure where to begin experimenting with this.

If you can offer some guidance on what code to input and a few things to try, I can begin experimenting with things tomorrow.

Also, thank you for the quick reply and your help!

**Comment by drJeckyll** [comment]
Well ... I expected to be something like that. API have limit of 1000 req/min. But why they put rate limit on frontend is beyond me.

OK. So remove that echo line which you put early. Then just bellow:
`$q = "https://www.tumblr.com" . $next;` on line 200 put:
`sleep(2);` to make delay between requests of 2 seconds. This may be will need to be adjusted, but you can try with different values.

**Comment by drJeckyll** [comment]
You can get new code with rate limit option from my gitlab: https://git.jeckyll.net/published/personal/tumblr-chat-messages-downloader

I will not update github repos anymore

**Comment by TheVeeC** [comment]
Ha, unfortunately I saw your post about the new code after I finally got it to work.  That would have probably been much easier.

I noticed the (higher) scroll-ups took about 4 seconds for loading new messages, so I used that instead of 2.  It worked, just took a few hours.  Managed to get the entire log finally!

But at least it worked!

I'll go ahead and close this issue.  Thank you again for your help and for coding this in the first place!
